### PR TITLE
Updating CFT spec with renaming and `CFTokenAuthorize` tx

### PR DESCRIPTION
**Pull request by shawnxie999**
-  Change the `CFTokenIssuanceID` construct to be issuer and the transaction sequence. 
- Removed AssetCode, now issuers should utilize metadata entirely
- Change freeze terminology to locking. And replaced `lsfCannotFreezeBalances` with `lsfCanLock`, which decides issuer's locking permission both individually and globally.
- Removed `Account` field in `CFTokenIssuanceSet`, this field is default to be issuer's address. We need another field called `CFTokenHolder` to specify the optional holder's account. 
- Renamed `Amount` field to `CFTokenAmount` in the `CFToken` object, because `Amount` is reserved for `STAmount` type only
- Added 1.7.0 section to describe how reserve requirements for CFTokenPages work.
- Added `CFTokenAuthorize` transaction
- New Appendix describing how allowlisting works for CFTs
- Transfer fee rounding behavior

### CODE REVIEW

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [inline XLS-0033d-CFTs/README.md:847]: Typo: accounts to `hold` their CFT.

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [inline XLS-0033d-CFTs/README.md:852]: Typo: Bob can now receive and ...

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [inline XLS-0033d-CFTs/README.md:854]: Type: will `successfully` delete.

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [inline XLS-0033d-CFTs/README.md:859]: Type: explore `the` flow and compare.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:86]: AssetCode is now only a metadata field. Each CFTokenIssuance is identified by only the ID, not the currency code.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: With this new proposal, AssetCode will become only metadata, which may or may not used. So now, the `Amount` field will need a change, as opposed to the original design: 
```
"amount": {
  "issuer": "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
  "currency": "USD",
  "value": "1"
}
```
It will now change to 
```
"amount": {
  "cft_issuance_id": "00070C4495F14B0E44F78A264E41713C64B5F89242540EE255534400000000000000",
  "value": "1"
}
```

After specifying the `CFTokenIssuanceID` in a `Payment` transaction, we are still able to fetch the `CFTokenIssuance` in constant time. 

Please let me know if I misunderstood anything

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: How is `cft_issuance_id` obtained?

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:86]: What is the motivation for this change?

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: Since the `CFTokenIssuanceID` will be the identifier, whoever wants to make a payment with that specific CFT needs to know the `CFTokenIssuanceID` before hand.

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: Yes, so how do they get this id? It's clear if I want to send say `USD`, but this is just pretty random identifier, which as I user I have to somehow locate.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: This would mean that we need to add an extra variable in the `Issue` class to encompass for the `CFTokenIssuanceID` in addition to  the `currency` and `issuer`

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: > Yes, so how do they get this id? It's clear if I want to send say USD, but this is just pretty random identifier, which as I user I have to somehow locate.

Yes, USD will no longer be an identifier. You, as the user, needs to keep track of the CFTokenIssuanceID, which is a random identifier. 

The reason why we changed to this approach because there is a problem where the issuer can re-issue the same asset with the same currency after it has been destroyed, which is a problem  for its use cases, according to product.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: This is why we are using the transaction sequence, which is a non-decreasing index that will uniquely identify a CFT

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: You are talking about implementation and it's not obvious why this should be added to `Issue` since the `currency` is irrelevant in this case. 
My question is how would a client application know the `CFTokenIssuanceID`?

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: Oh, so we are completely relying on clio for API support. 

There are three APIs:

`cfts_by_issuer`: lists all the CFTokenIssuanceID issued by the account
`account_cfts`: all the CFT balances of  an account
`cft_holders`: All the accounts who hold this particular

**Comment by gregtatcam** [comment]
Should the `1.1.2. Assumptions` section be updated since we are now introducing essentially `TrustSet` transaction?

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: I'm talking about the payment transaction and looking forward offer create, amm create, etc.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: But it's definitely less convinient for the user since they can no longer specify the currency directly

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: Yes, it is really inconvenient. Is there a PRD or other relevant documentation, which discusses the motivation for this change?

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: I think we can discuss it later in a call.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:168]: But is there a technical limitation with the new approach? I have done too much with payment or offer so i dont really know.

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:168]: I have not had time to give it a lot of thought. But it seems like on the surface it's a big change.

File: XLS-0033d-CFTs/README.md
Comment (gregtatcam) [inline XLS-0033d-CFTs/README.md:902]: Could you please clarify "This behavior can be misleading to token holders who have held onto the first iteration..."? `CFTokenIssuance` can only be destroyed if there are no holders of CFT so once it's destroyed, there is tokens left from this iteration.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:902]: Yeah I'm not too exactly sure of this either...I think Shashwat might have better idea from the product perspective.

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:902]: And this is not really different in that respect from IOU. I.e. the issuer can delete the trustline once all IOU's are redeemed and then create it again. It would be good to understand what problem we are really solving by going the uniqueid approach for each new `CFTokenIssuance` rather than using issuer/assetCode for the id.

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:902]: i agree

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [inline XLS-0033d-CFTs/README.md:848]: Let's move this to an Appendix. This is not normative for this spec.

**Review by sappenin — CHANGES_REQUESTED** [review]

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:191]: Good point!

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:362]: This is for `CFTokenIssuanceCreate` transaction, so it's fine if the issuer doesn't specify it. If `AssetScale` is not specified, the ledger will automatically default it to zero.

Also, btw, `AssetScale` will not involved in on-ledger math at all _I think_. It is merely a field used for metadata as well. 
So assume the issuer has set USD to use `AssetScale` of 2. It means USD will be dealt in basic unit of a cent. So if you specify a payment tx like:
```
"amount": {
  "issuer": "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
  "cft_asset": "USD",
  "value": "150"
}
```
This means that the account wants to send $1.50 USD. But on the ledger, math is still going to be done as 150 whole units.

Comment (gregtatcam) [reply XLS-0033d-CFTs/README.md:362]: Correct. I think `AssetScale` is only needed for display purpose.

Comment (sappenin) [reply XLS-0033d-CFTs/README.md:362]: > AssetScale is only needed for display purpose.

I think this is correct, but for the avoidance of doubt, what we're really saying is that all math related to CFTs, and all transaction inputs, will be done with an implied scale of 0 (sort of like the way we operate on drops of XRP).

Agree?

Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:362]: Yeah, when doing math for CFT, what the `AssetScale` is(whether it's 0 or 100), is not relevant at all

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:844]: It's a product requirement, we would want to incentive holders to use CFT, also less learning curve, as they don't need to care about reserve requirements. Giving them free CFTokenPage would mean that they can get 32 CFTs free of reserves

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:559]: If the holder calls `CFTokenTrust` with the `tfUntrust` flag, it will delete holder's CFToken as long as the amount is zero. Are you talking about introducing a new `CFTokenDelete` transaction? Because we currently don't have transaction that's called this.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md]: > Do we need both Account and CFTokenHolder fields in this CFTokenTrust transaction?

`Account` field is a mandatory field in transaction I believe. `CFTokenHolder` will always be holder's address, it will only be used when issuer wants to authorize a holder.

> As an alternative design: If we only used `CFTokenHolder`, then this could always specify what it says -- the holder address. Here we can consider the following scenarios:
> 
> 1. Issuer is authorizing a holder (single-sig): The transactor looks-up the `CFTokenIssuanceID` and finds the issuer address from there (i.e., the `Issuer` field). Signed by the issuer.
> 2. Issuer is authorizing a holder (multi-sig): Same as above (signed by issuer quorum)
> 3. Holder is authorizing an issuer (single-sig): Same as above (must be signed by holder)
> 4. Holder is authorizing an issuer (multi-sig): Same as above (signed by holder quorum).

Allowlisting has nothing to do with with multisigning (unless you are suggesting otherwise)

To clear things up, there are two main scenario how CFTokenTrust will be used:

### Example
Assume that Alice is the issuer of USD, and Bob wants to use it.

#### If USD doesn't have allowlisting:
1. Bob sends a CFTokenTrust to hold Alice's token like:
```
{
  TransactionType: "CFTokenTrust",
  Account: Bob,
  CFTokenIssuanceID: "123",
}
```
This creates a zero-balance `CFToken` object for Bob, and now Bob can receive USD from anybody. Note here, that `CFTokenHolder` is not specified, because Bob's address(holder) is in the `Account` field

#### If USD has allowlisting enabled:
1. (Same as above)Bob sends a CFTokenTrust to hold Alice's token like:
```
{
  TransactionType: "CFTokenTrust",
  Account: Bob,
  CFTokenIssuanceID: "123",
}
```

2. Alice(issuer) sends a CFTokenTrust to authorize Bob:
```
{
  TransactionType: "CFTokenTrust",
  Account: Alice,
  CFTokenIssuanceID: "123",
  CFTokenHolder: Bob
}
```
If this transaction is successful, it will set the `tfAuthorized` flag on Bob's CFToken.
Here, we need both `Account` and `CFTokenHolder` because Alice needs to specify which holder's account she wants to authorize, in case, she want's to authorize Bob to use USD.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md]: Fixed. Removed it completely

File: XLS-0033d-CFTs/README.md
Comment (kennyzlei) [reply XLS-0033d-CFTs/README.md:118]: Ethereum has the equivalent concept of "Decimals" on chain, and I find some value in keeping that

https://eips.ethereum.org/EIPS/eip-20

Comment (sappenin) [reply XLS-0033d-CFTs/README.md:118]: Not opposed - I also find some value of it on-chain. However, I was thinking it would just be located in `CFTokenMetadata` (but still on-chain).

Are you saying you prefer `AssetScale` _outside_ of metadata?

Comment (kennyzlei) [reply XLS-0033d-CFTs/README.md:118]: Yeah I am advocating for it be a first-class field outside of metadata, although I can understand the rationale that everything off-chain purpose should live in metadata.  I think it comes down to, "should off-chain fields that we want standardization on live in metadata and hope people adopt, or should we enforce it using a first-class field?"  Either outcome is reasonable to me though, no strong opinion here

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:510]: The transaction has already prefixed with CFToken so I believe the confusion will be minimal. 

I believe "trust" is the right term here, since the holder still need to trust the issuer in order to redeem the tokens, so I don't really see this word being exclusive to trustlines. I'm open to renaming this transaction but right now I can't really think of better term that could replace it.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md:118]: I think we can leave it as it is for now(having it as a separate field). It is something that is fairly minor and easy to change, so we can make modifications anytime during implementation phase.

File: XLS-0033d-CFTs/README.md
Comment (shawnxie999) [reply XLS-0033d-CFTs/README.md]: Actually I changed my mind. I am looking to change it to `CFTokenAuthorize`, I do think this is a better naming, especially when there is allow-listing involved.

File: XLS-0033d-compact-fungible-tokens/README.md
Comment (sappenin) [inline XLS-0033d-compact-fungible-tokens/README.md:208]:
```suggestion
| `lsfAuthorized`       | `0x0002`   | (Only applicable for allow-listing) If set, indicates that the issuer has authorized the holder for the CFT. This flag can be set using a `CFTokenAuthorize` transaction; it can also be "un-set" using a `CFTokenAuthorize` transaction specifying the `tfUnauthorize` flag. |
```

Comment (sappenin) [reply XLS-0033d-compact-fungible-tokens/README.md:208]: There was a minor typo in the original, but I decided to suggest a few more edits.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: For posterity, this change was made via 270356c03f8e76e4ce4de002a8732bb6f8f122d2

File: XLS-0033d-compact-fungible-tokens/README.md
Comment (sappenin) [inline XLS-0033d-compact-fungible-tokens/README.md:350]:
```suggestion
An asset scale is the difference, in orders of magnitude, between a standard unit and a corresponding fractional unit. More formally, the asset scale is a non-negative integer (0, 1, 2, …) such that one standard unit equals 10^(-scale) of a corresponding fractional unit. If the fractional unit equals the standard unit, then the asset scale is 0. Note that this value is optional, and will default to `0` if not supplied.
```

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: Resolving this for now. I'm comfortable making this optional here -- I'll likely submit a future PR to remove this field entirely, but we can debate that in the future, in a separate PR.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: I'm not sure - I'll revisit this idea later once this PR is merged.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: I don't love this "requirement" for a few reasons:

1. It shields users from reserve requirements, which is likely to cause confusion -- once a user exceeds whatever arbitrary "free" limit we give them -- but only later after a user has been using CFTs for awhile. I would prefer that users encounter reserve fees right away, because this is something that users probably need to be aware of anyway.
2. Reserve fees exist for a reason, which is to pay for space on-ledger. There's a decent chance that a large group of people hold fewer than 32 CFTs, which means there's a decent chance that a large group of people make _many_ CFTs and none of them pay reserve fees. Put another way, I worry that this "requirement" essentially exempts CFTs from reserve fees, which seems counter-productive (assuming reserve fees are there for a reasonable purpose, which I believe to be true).
3. Last but not least, this likely introduces unnecessary code/implementation complexity around ledger fee computation.

That said, I'm going to resolve this thread to unblock this PR. We can (potentially) debate this in another PR.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: Sounds good to me to revisit later. Resolving for now, and will open a PR potentially in the future to revisit.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: I will introduce a separate PR to define a JSON schema, and we can address this question separately.

**Review by sappenin — APPROVED** [review]
LGTM - approving to unblock this current round of changes, but for anyone monitoring, expect more updates to come. For example, I'll create some more PRs with discrete items so we don't have a massive change-set moving forward.

File: XLS-0033d-CFTs/README.md
Comment (sappenin) [reply XLS-0033d-CFTs/README.md]: Resolving this to unblock for now. I'll open another PR after I think through this some more (to decided if it's still an issue).

**Pull request merged**
